src/Microsoft.OData.Core/JsonLight/ODataJsonLightResourceDeserializer.cs

```diff
@@ -1061,136 +1062,2239 @@ internal ODataJsonLightReaderNestedInfo ReadPropertyWithoutValue(IODataJsonLight
+        ///                 JsonNodeType.PrimitiveValue  if the resource set is a member of an untyped collection followed by a primitive value
+        ///                 JsonNodeType.StartObject     if the resource set is a member of an untyped collection followed by a resource
+        ///                 JsonNodeType.EndArray        if the resource set is the last member of an untyped collection
+        /// </remarks>
+        internal async Task ReadResourceSetContentEndAsync()
+        {
+            this.AssertJsonCondition(JsonNodeType.EndArray);
+            this.JsonReader.AssertNotBuffering();
+
+            await this.JsonReader.ReadEndArrayAsync()
+                .ConfigureAwait(false);
+
+            this.AssertJsonCondition(
+                JsonNodeType.EndOfInput,
+                JsonNodeType.EndObject,
+                JsonNodeType.Property,
+                JsonNodeType.StartArray,
+                JsonNodeType.PrimitiveValue,
+                JsonNodeType.StartObject,
+                JsonNodeType.EndArray);
+        }
+
+        /// <summary>
+        /// Asynchronously reads the resource type name annotation (odata.type)
+        /// </summary>
+        /// <returns>A task that represents the asynchronous read operation.</returns>
+        /// <param name="resourceState">The state of the reader for resource to read.</param>
+        /// <remarks>
+        /// Pre-Condition:  JsonNodeType.Property          The first property after the odata.context in the resource object.
+        ///                 JsonNodeType.EndObject         End of the resource object.
+        /// Post-Condition: JsonNodeType.Property          The property after the odata.type (if there was any), or the property on which the method was called.
+        ///                 JsonNodeType.EndObject         End of the resource object.
+        ///
+        /// This method fills the ODataResource.TypeName property if the type name is found in the payload.
+        /// </remarks>
+        internal async Task ReadResourceTypeNameAsync(IODataJsonLightReaderResourceState resourceState)
+        {
+            Debug.Assert(resourceState != null, "resourceState != null");
+            this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
+
+            // If the current node is the odata.type property - read it.
+            if (this.JsonReader.NodeType == JsonNodeType.Property)
+            {
+                string propertyName = await this.JsonReader.GetPropertyNameAsync()
+                    .ConfigureAwait(false);
+                // TODO: const to avoid repeated concatenation - JsonLightConstants.ODataPropertyAnnotationSeparatorChar + ODataAnnotationNames.ODataType?
+                if (string.CompareOrdinal(JsonLightConstants.ODataPropertyAnnotationSeparatorChar + ODataAnnotationNames.ODataType, propertyName) == 0

```

Comment:
nit:
```suggestion
                if (string.CompareOrdinal(string.Concat(JsonLightConstants.ODataPropertyAnnotationSeparatorChar, ODataAnnotationNames.ODataType), propertyName) == 0

```

Reply:
I am guessing the compile will do this as well behind the scenes.

Reply:
@marabooy Fixed by introducing constants comprising of the concatenated string